File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;
vector<vector<int>> graph;
vector<bool> visit;
int cnt =0; 
void dfs(int);
int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    visit.resize(n, false); 
    graph = computers; 

    for(int i=0 ;i<n;i++) {
        cnt =0; 
        dfs(i);
        if(cnt>0) {
            answer++; 
        }
    }

    return answer;
}

void dfs(int n) {
    if(visit[n]) {
        return; 
    }
    
    visit[n]=true;
    cnt++; 
    
    for(int i=0; i<graph[n].size(); i++) {
        if(graph[n][i]==1&& visit[i]==false)  {
            dfs(i);
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    ArrayList<ArrayList<Integer>> network;
    boolean[] visited;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        network = new ArrayList<>();
        for(int i=0 ;i<n; i++){
            network.add(new ArrayList<>());
        }
        visited = new boolean[n];
        for(int i=0; i<computers.length; i++){
            for(int j=0; j<computers[0].length; j++){
                if(i==j){continue;}
                if(computers[i][j]==1){
                    network.get(i).add(j);
                }
            }
        }
        
        for(int i=0; i<n; i++){
            if(!visited[i]){
                dfs(i);
                answer++;
                System.out.println(i);
            }
        }
        return answer;
    }
    
    void dfs(int cn){
        if(visited[cn]){
            return;
        }
        visited[cn]=true;
        for(int i=0; i<network.get(cn).size(); i++){
            if(!visited[network.get(cn).get(i)]){
                dfs(network.get(cn).get(i));
            }
        }
    }
}